File: data-structures-algorithms/leetcode/309.go
func getMaxNum(num1 int, num2 int) int {
	if num1 > num2 {
		return num1
	}
	return num2
}

func maxProfit(prices []int) int {
	length := len(prices)
	if length == 1 {
		return 0
	}
	memo := make([][3]int, length)
	memo[0][0] = -prices[0]

	for i := 1; i < length; i++ {
		// 有股票时，最大收益
		memo[i][0] = getMaxNum(memo[i-1][0], memo[i-1][2]-prices[i])
		// 没有股票且在冷冻期，最大收益
		memo[i][1] = memo[i-1][0] + prices[i]
		// 没有股票且不在冷冻期，最大收益
		memo[i][2] = getMaxNum(memo[i-1][1], memo[i-1][2])
	}

	return getMaxNum(memo[length-1][1], memo[length-1][2])
}

File: data-structures-algorithms/leetcode/js/309.js
var maxProfit = function(prices) {
  const len = prices.length
  
  const maxProfitDp = new Array(len).fill(0).map(() => new Array(3).fill(0))
  // 有股票的最大收益
  maxProfitDp[0][0] = -prices[0]
  for (let i = 1; i < len; i++) {
    maxProfitDp[i][0] = Math.max(maxProfitDp[i - 1][0], maxProfitDp[i - 1][2] - prices[i])
    // 没有股票在冷冻期的最大收益
    maxProfitDp[i][1] = maxProfitDp[i - 1][0] + prices[i]
    // 没有股票不在冷冻期的最大收益
    maxProfitDp[i][2] = Math.max(maxProfitDp[i - 1][2], maxProfitDp[i - 1][1])
  }

  return Math.max(maxProfitDp[len - 1][1], maxProfitDp[len - 1][2])
}
